Did you test it? For some reason it seems to fail.
```
    _ = special_math_ops.einsum('ii->', x)
  File "/usr/lib/python3.4/unittest/case.py", line 181, in __exit__
    self._raiseFailure("{} not raised".format(exc_name))
  File "/usr/lib/python3.4/unittest/case.py", line 130, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: ValueError not raised
```

--- a/tensorflow/python/ops/special_math_ops.py
+++ b/tensorflow/python/ops/special_math_ops.py
@@ -240,7 +240,7 @@ def einsum(equation, *inputs, **kwargs):
     for a in axis_labels:
       for input_labels in input_axis_labels:
         if (len(input_axis_labels) == 1 and input_labels.count(a) == 2
-            and input_labels == input_labels[::-1]):
+            and input_labels == input_labels[::-1] and '->' not in equation):
           return math_ops.trace(inputs[0])
         if input_labels.count(a) > 1:
           raise ValueError(

